Unify build system

WalterKolczynski-NOAA opened the pull request:
Combines the checkout and build from the standard and coupled workflows
system that checks out the current UFS.

The model source directory is now ufs_model.fd rather than fv3gfs.fd or
ufs_coupled.fd. The build script is renamed build_ufs.sh and the link
script is renamed to simply link.sh.

The checkout script checks out the same UFS version (current as of
2021 Jul 20) for all applications, removing the coupled option from the
coupled branch. A new option, -m, allows the user to specify a different
model hash to check out. Checkout log files are now placed in the logs
directory like the build logs in order to keep the sorc directory clean.

Two new options are added to the build_all script: -c to build the
coupled (S2SW) app, and -a to build the aerosol (ATMAERO) app. Without
either argument, the traditional (ATMW) app is built. If either option
is provided, the cpl_build.cfg file controls which executables are built;
otherwise fv3gfs_build.cfg is still used. The build_ufs script corre-
spondingly accepts the same options.

The reg2grb2 source directory is copied over from the coupled model and
cmake files for workflow_utils are updated correspondingly to build.
Additionally, the build_workflow_utils modulefiles are updated to use the
ncio module in hpc-stack and the hack that cloned and built it as part of
the workflow_utils build is removed.

The build_ww3prepost script is updated to make the code more easily main-
tained and the corresponding modulefiles are updated to remove unnecessary
modules.

The link script is updated to the new fix location (ending in fix_NEW)
removes links with targets that no longer exist. A new third argument to
link, coupled, is added to additionally link files needed by the coupled
system.

Note that while the checkout, build, and link have all been tested and
complete without error, the workflow is otherwise untested and may not
work until additional updates are made to use the current UFS.

Refs: #366, #368, #374

WalterKolczynski-NOAA committed:
Unify build system

Combines the checkout and build from the standard and coupled workflows
system that checks out the current UFS.

The model source directory is now ufs_model.fd rather than fv3gfs.fd or
ufs_coupled.fd. The build script is renamed build_ufs.sh and the link
script is renamed to simply link.sh.

The checkout script checks out the same UFS version (current as of
2021 Jul 20) for all applications, removing the coupled option from the
coupled branch. A new option, -m, allows the user to specify a different
model hash to check out. Checkout log files are now placed in the logs
directory like the build logs in order to keep the sorc directory clean.

Two new options are added to the build_all script: -c to build the
coupled (S2SW) app, and -a to build the aerosol (ATMAERO) app. Without
either argument, the traditional (ATMW) app is built. If either option
is provided, the cpl_build.cfg file controls which executables are built;
otherwise fv3gfs_build.cfg is still used. The build_ufs script corre-
spondingly accepts the same options.

The reg2grb2 source directory is copied over from the coupled model and
cmake files for workflow_utils are updated correspondingly to build.
Additionally, the build_workflow_utils modulefiles are updated to use the
ncio module in hpc-stack and the hack that cloned and built it as part of
the workflow_utils build is removed.

The build_ww3prepost script is updated to make the code more easily main-
tained and the corresponding modulefiles are updated to remove unnecessary
modules.

The link script is updated to the new fix location (ending in fix_NEW)
removes links with targets that no longer exist. A new third argument to
link, coupled, is added to additionally link files needed by the coupled
system.

Note that while the checkout, build, and link have all been tested and
complete without error, the workflow is otherwise untested and may not
work until additional updates are made to use the current UFS.

Refs: #366, #368, #374

aerorahul commented on sorc/build_ufs.sh:
```diff
@@ -0,0 +1,49 @@
+cd ufs_model.fd/
+module use ${MOD_PATH}
+module load ufs_${target}
+CMAKE_FLAGS="-DAPP=${APP} -DCCPP_SUITES=${CCPP_SUITES}" ./build.sh
```
Please don't use the `build.sh` in the UFS weather model.  Its use is for the regression tests.  Users and developers should use `cmake` instructions required to build the model.

```suggestion
mkdir -p build && cd build
cmake -DAPP=${APP} -DCCPP_SUITES=${CCPP_SUITES} ..
OMP_NUM_THREADS=1 make -j ${BUILD_JOBS:-8} VERBOSE=${BUILD_VERBOSE:-} 
```

WalterKolczynski-NOAA replied:
I am fine making this change, but until now this is how we've been building UFS in both branches.

aerorahul replied:
Sure. 
Going forward, if we want to maintain portability.  We should:
- ensure we use the build instructions and not the build script provided by the components
- minimize reliance on modulefiles (this is impossible on supported platforms).  But, if we use build instructions, we take the module load environment out of the equation, since the module files provide paths to the third party libraries.  So, if we land on a machine with no modules but the build environment in the `PATH`, we can build without modulefiles and the component build scripts.

WalterKolczynski-NOAA replied:
Updated to use cmake directly. Per conversation, postponed moving build directory out of UFS repo. A future PR will do that for all components.

aerorahul commented on sorc/link.sh:
```diff
@@ -5,19 +5,30 @@ set -ex
 
```
I don't think just `link.sh` is appropriate as it is too generic.
Please retain a suffix and ensure it is consistent with the `usage` blurb inside the script.

WalterKolczynski-NOAA replied:
Updated to link_workflow.sh

WalterKolczynski-NOAA committed:
Rename link script, update usage

The link script is updated to something indicating it is in the
workflow.

Link script usage statement is updated to reflect its new name.

Refs: #366, #368, #374

KateFriedman-NOAA commented on sorc/checkout.sh:
```diff
@@ -21,24 +24,24 @@ do
 else
-    echo 'Skip.  Directory fv3gfs.fd already exists.'
-fi
+    echo 'Skip.  Directory ufs.fd already exists.'
```
Should be "ufs_model.fd already exists."

WalterKolczynski-NOAA replied:
Yes, it should. Fixed.

KateFriedman-NOAA commented on sorc/fv3gfs_build.cfg:
```diff
@@ -1,7 +1,7 @@
 # ***** configuration of global-workflow build *****
 
- Building fv3gfs (fv3gfs) .............................. yes
+ Building ufs_model (ufs_model) ........................ yes
```
Why retain this build config if we have cpl_build.cfg now?

WalterKolczynski-NOAA replied:
The coupled config doesn't build most of the parts used by GFS (like GSI). Eventually it will, but until then there is no use building parts not needed for the prototypes. There probably should be a separate one for aerosols too (they don't need WW3 pre/post, for example).

aerorahul replied:
Should this file be renamed from `fv3gfs_build.cfg` to `gfs_build.cfg`. I assume `fv3gfs` refers to the application and not just the fv3-based model.

aerorahul replied:
looking at the `link_fv3gfs.sh` file renamed below, another alternative name `workflow_build.cfg`?
Just throwing out suggestions.

WalterKolczynski-NOAA replied:
Yeah, in the future we can rename these files so they are consistent/clearer. We can worry about that later though.

WalterKolczynski-NOAA committed:
Fix exists message for UFS model in checkout script

The directory exists message for the UFS model erroneously stated
the ufs.fd directory exists instead of the ufs_model.fd directory
exists.

Refs: #366, #368, #374

WalterKolczynski-NOAA committed:
Use cmake directly in UFS build

The UFS build script now calls cmake directly instead of using the
build script located in UFS.

In the future, the build directory for all components will be moved
out of the cloned repos and into workflow.

Refs: #366, #368, #374

JessicaMeixner-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -68,7 +100,11 @@ cd ${pwd}/../parm               ||exit 8
     $LINK ../sorc/gfs_post.fd/scripts/exgdas_atmos_nceppost.sh .
-    $LINK ../sorc/gfs_post.fd/scripts/exgfs_atmos_nceppost.sh  .
+    if [ $model = "coupled" ]; then
+      $LINK exgfs_nceppost_cpl.sh exgfs_atmos_nceppost.sh
```
This seems to be the last special section for "coupled", I forget what the issue/difference is here, but should we open an issue(s) somewhere to get this resolved so we don't have to have these two different copies?

JessicaMeixner-NOAA replied:
Well, there's one other "coupled" section, but comment still remains.

WalterKolczynski-NOAA replied:
I expect there will be a bunch of little things like this to hammer out as we determine which is because newer UFS and which is because coupled.

WalterKolczynski-NOAA replied:
This script in particular is the one Arun so much wants to move out of post and into workflow anyway (and replace it with one that just does the regridding now that we use inline post).

KateFriedman-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -207,29 +247,29 @@ for workflowexec in fbwndgfs gfs_bufr regrid_nemsio supvit syndat_getjtbul \
   $LINK ../sorc/install/bin/${workflowexec}.x $workflowexec
 done
 for workflowexec in enkf_chgres_recenter.x enkf_chgres_recenter_nc.x fv3nc2nemsio.x \
-    tave.x vint.x ; do
+    tave.x vint.x reg2grb2.x ; do
   [[ -s $workflowexec ]] && rm -f $workflowexec
   $LINK ../sorc/install/bin/$workflowexec .
 done
 
-[[ -s global_fv3gfs.x ]] && rm -f global_fv3gfs.x
-$LINK ../sorc/fv3gfs.fd/NEMS/exe/global_fv3gfs.x .
+[[ -s ufs_model ]] && rm -f ufs_model
+$LINK ../sorc/ufs_model.fd/build/ufs_model .
 
 [[ -s gfs_ncep_post ]] && rm -f gfs_ncep_post
 $LINK ../sorc/gfs_post.fd/exec/upp.x gfs_ncep_post
 
 if [ -d ${pwd}/gfs_wafs.fd ]; then 
     for wafsexe in \
-          wafs_awc_wafavn.x  wafs_blending.x  wafs_blending_0p25.x \
-          wafs_cnvgrib2.x  wafs_gcip.x  wafs_grib2_0p25.x \
-          wafs_makewafs.x  wafs_setmissing.x ; do
+          wafs_awc_wafavn  wafs_blending  wafs_blending_0p25 \
+          wafs_cnvgrib2  wafs_gcip  wafs_grib2_0p25 \
+          wafs_makewafs  wafs_setmissing; do
```
The hash of WAFS in the develop branch produces execs with ".x" in the name. Does the linking work with this change?

WalterKolczynski-NOAA replied:
I'm going to guess no. I didn't pick up on this in testing because no WAFS links are produced at all, so there were no broken links to see.

KateFriedman-NOAA reviewed (changes requested):
Please revert removal of ".x" from WAFS exec linking. Changes look good otherwise, thanks!

WalterKolczynski-NOAA committed:
Fix linked executables for WAFS

The executables for WAFS have a .x suffix. The link script now uses
the correct executable names.

Refs: #366, #368, #374

JessicaMeixner-NOAA commented on sorc/checkout.sh:
```diff
@@ -72,11 +75,10 @@ fi
+    git clone https://github.com/NOAA-EMC/EMC_post.git gfs_post.fd >> ${logdir}/checkout-gfs_post.log 2>&1
     cd gfs_post.fd
-    git checkout ff42e0227d6100285d4179a2572b700fd5a959cb
+    git checkout upp_v10.0.3
```
https://github.com/NOAA-EMC/EMC_post/commit/ff42e0227d6100285d4179a2572b700fd5a959cb is from April where as the upp_v10.0.3 is from January.  Should we use the more recent UPP hash?

KateFriedman-NOAA replied:
If the coupled system doesn't specifically need the upp_v10.0.3 tag then I agree that the newer hash should remain. That's the hash that was tested with the hpc-stack updates commit.

WalterKolczynski-NOAA replied:
Should we just use the latest UPP tag here?

aerorahul replied:
the atmosphere only does inline post, so I don't think that will break @JianKuang-UMD's branch.

WalterKolczynski-NOAA replied:
module_base is loading upp/10.0.6, so I propose we check out that version

WalterKolczynski-NOAA replied:
Alternatively, UPP is now up to 10.0.8. I can update all of them to that tag

aerorahul replied:
sure.  I don't really have a preference.  It is used for offline post processing, and who is using that now?

WalterKolczynski-NOAA replied:
I think we might be using some of the scripts? But I don't think the version really matters all that much.

WalterKolczynski-NOAA replied:
Updated all to 10.0.8 for consistency

aerorahul commented on modulefiles/modulefile.ww3.wcoss_dell_p3:
```diff
@@ -9,20 +9,7 @@ module load jasper/2.0.25
-module load wgrib2/2.0.8
-module load upp/10.0.6
+
+module load netcdf/4.7.4
```
we should load the `netcdf` dependency `hdf5` before netcdf.

aerorahul replied:
Also, move this up before any other module load that depends on these, e.g. `upp`

WalterKolczynski-NOAA replied:
UPP is no longer loaded for this. The only other modules loaded are the three compression libraries, bacio, and g2.

aerorahul commented on modulefiles/modulefile.ww3.orion:
```diff
@@ -14,20 +14,7 @@ module load jasper/2.0.25
-module load wgrib2/2.0.8
-module load upp/10.0.6
+
+module load netcdf/4.7.4
```
load `hdf5` dependency explicitly before `netcdf`.

WalterKolczynski-NOAA replied:
Done

aerorahul commented on modulefiles/workflow_utils.hera:
```diff
@@ -20,7 +20,9 @@ module load sp/2.3.3
 module load nemsio/2.5.2
 module load nemsiogfs/2.5.3
-#module load ncio/1.0.0
+module load ncio/1.0.0
```
Please move `hdf5` and `netcdf` from lines 30-31 above.
`ncio` depends on `netcdf` and order matters.
We get away with it here because we build everything as static libraries.

WalterKolczynski-NOAA replied:
Done

aerorahul replied:
I think w3emc/2.7.3 depends on netcdf.
Later versions of w3emc removed that dependency.  These should be moved before the nceplibs list start since these are more closer to third-party than the nceplibs ones.

aerorahul commented on modulefiles/workflow_utils.orion:
```diff
@@ -20,7 +20,9 @@ module load sp/2.3.3
 module load nemsio/2.5.2
 module load nemsiogfs/2.5.3
-#module load ncio/1.0.0
+module load ncio/1.0.0
```
same comment as for hera

WalterKolczynski-NOAA replied:
Done

aerorahul replied:
same comment as in hera.

aerorahul commented on sorc/build_all.sh:
```diff
@@ -9,6 +9,25 @@ set +x
 #                   Anything other than "true"  will use libraries locally.
 #------------------------------------
 
+_build_ufs_options=""
+
+while getopts "ac" option;do
+ case $option in
+  a)
+   echo "Received -a flag, building ufs-weather-model for ATMAERO app"
+   echo "skipping builds not needed for prototype runs"
+   _build_ufs_options=-a
+   break
+   ;;
+  c)
+   echo "Received -c flag, building ufs-weather-model for S2SW app"
+   echo "skipping builds not needed for prototype runs"
+   _build_ufs_options=-c
+   break
+   ;;
+ esac
+done
```
We will have to revisit this approach in the future.
Imagine if every component covered in `build_all.sh` has options to build, this argument list will get unmanageable.

WalterKolczynski-NOAA replied:
Don't disagree. Right now this is just a shortcut for the APP to build. We could easily just change this to take the app as an argument instead of these switches. Although, hopefully in the near future we can build one app executable and use it for all configurations (at least until regional gets brought in or something, by which time this will be replaced).

aerorahul commented on sorc/build_ufs.sh:
```diff
@@ -0,0 +1,52 @@
+
+# Set target platform
+case "${target}" in
+  hera|orion|stampede)
```
might as well throw `jet` and `cheyenne` here.  They follow the same pattern.

aerorahul replied:
```suggestion
  hera|orion|stampede|jet|cheyenne)
```

aerorahul commented on sorc/build_ufs.sh:
```diff
@@ -0,0 +1,52 @@
+module purge
+
+cd ufs_model.fd/
+module use ${MOD_PATH}
+module load ufs_${target}
```
please wrap all `module` commands in `set +x` and `set -x`.

WalterKolczynski-NOAA replied:
Done

aerorahul commented on sorc/build_ww3prepost.sh:
```diff
@@ -35,6 +34,7 @@ export Z_LIB=${Z_LIB:-$ZLIB_ROOT/lib/libz.a}
 export JASPER_LIB=${JASPER_LIB:-$JASPER_ROOT/lib64/libjasper.a}
 export WWATCH3_NETCDF=NC4
 export NETCDF_CONFIG=$NETCDF_ROOT/bin/nc-config
+export PRINTER=${PRINTER:-""}
```
I wonder what is `PRINTER` doing?

JessicaMeixner-NOAA replied:
it's a variable that's defined as part of the ww3 build which is getting cleaned out.  So this should go away soon.  I'm assuming Walter added it because a version he was using needed it for some reason?

WalterKolczynski-NOAA replied:
I turned on set -u and the script errored because it wasn't defined. So I defined it.

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -30,18 +41,39 @@ pwd=$(pwd -P)
-    FIX_DIR="/work/noaa/global/glopara/fix"
+    FIX_DIR="/work/noaa/global/glopara/fix_NEW"
+elif [ $machine = "stampede" ]; then
+    FIX_DIR="/work/07738/jkuang/stampede2/tempFixICdir/fix_UFSp6"
```
This is pointing to `jkuang`'s personal space.  Does this person know not to delete this path?

JessicaMeixner-NOAA replied:
We shouldn't even be pointing to that fix directory anymore b/c it's now out of date compared to the fix_NEW

WalterKolczynski-NOAA replied:
Stampede users have been left to fend for themselves for the time being. The stampede blocks were brought in because it had partially been added to the coupled branch and I just copied it. I can remove them all, though I am inclined to leave them for at least a little while in case we wind up returning to that effort. But I also don't see a need to update paths until that happens.

aerorahul replied:
Leave for now.

aerorahul reviewed (commented):
Just a few minor comments.  Otherwise looks good.

WalterKolczynski-NOAA committed:
Reorganize some module loads

The order of module loads is reordered so HDF and NetCDF are loaded
before modules that use them.

Refs: #366, #374

WalterKolczynski-NOAA committed:
Quiet UFS build module load

Module loading in the UFS build script now sets +x first to turn off
the command trace.

Refs: #366, #368, #374

aerorahul commented on sorc/build_ufs.sh:
```diff
@@ -0,0 +1,54 @@
+
+MOD_PATH=$cwd/ufs_model.fd/modulefiles
+
+module purge
```
please move this inside the `set +x` `set -x`

aerorahul replied:
```suggestion
```

aerorahul committed:
Update sorc/build_ufs.sh

aerorahul commented on sorc/build_ufs.sh:
```diff
@@ -0,0 +1,54 @@
+
+cd ufs_model.fd/
+set +x
+module use ${MOD_PATH}
```
```suggestion
module purge
module use ${MOD_PATH}
```

WalterKolczynski-NOAA committed:
Reorder HDF/NetCDF load order

Moves the HDF and NetCDF module loads before all the NCEP libs.

Refs: #366, #374

WalterKolczynski-NOAA committed:
Quiet UFS build module purge

Moved the UFS build module purge inside the set +x block.

Refs: #366, #374

WalterKolczynski-NOAA committed:
Update UPP to 10.0.8

Updates UPP to 10.0.8 to be consistent with UFS. This also adds
consistency between the moduleload and checked out version.

Refs: #366, #374
